cpp/src/io/csv/csv_reader.cu

```diff
@@ -686,45 +689,50 @@ gdf_error read_csv(csv_read_arg *args)
 	free(h_dtypes); 
 	free(h_valid); 
 	free(h_data); 
-	
-	launch_dataConvertColumns(raw_csv, d_data, d_valid, d_dtypes, d_str_cols, d_valid_count);
-	cudaDeviceSynchronize();
 
-	stringColCount=0;
-	for (int col = 0; col < raw_csv->num_active_cols; col++) {
+	if (raw_csv->num_records != 0) {
+		error = launch_dataConvertColumns(raw_csv, d_data, d_valid, d_dtypes, d_str_cols, d_valid_count);
+		if (error != GDF_SUCCESS) {
+			return error;
+		}
+		cudaDeviceSynchronize();
 
-		gdf_column *gdf = cols[col];
+		stringColCount=0;
+		for (int col = 0; col < raw_csv->num_active_cols; col++) {
 
-		if (gdf->dtype != gdf_dtype::GDF_STRING)
-			continue;
+			gdf_column *gdf = cols[col];
 
-		NVStrings* const stringCol = NVStrings::create_from_index(h_str_cols[stringColCount],size_t(raw_csv->num_records));
-		if ((raw_csv->quotechar != '\0') && (raw_csv->doublequote==true)) {
-			// In PANDAS, default of enabling doublequote for two consecutive
-			// quotechar in quote fields results in reduction to single
-			std::string quotechar = std::string(&raw_csv->quotechar);
-			std::string doublequotechar = quotechar + raw_csv->quotechar;
-			gdf->data = stringCol->replace(doublequotechar.c_str(), quotechar.c_str());
-			NVStrings::destroy(stringCol);
-		}
-		else {
-			gdf->data = stringCol;
-		}
+			if (gdf->dtype != gdf_dtype::GDF_STRING)
+				continue;
 
-		RMM_TRY( RMM_FREE( h_str_cols [stringColCount], 0 ) );
+			NVStrings* const stringCol = NVStrings::create_from_index(h_str_cols[stringColCount],size_t(raw_csv->num_records));
+			if ((raw_csv->quotechar != '\0') && (raw_csv->doublequote==true)) {
+				// In PANDAS, default of enabling doublequote for two consecutive
+				// quotechar in quote fields results in reduction to single
+				std::string quotechar = std::string(&raw_csv->quotechar);
+				std::string doublequotechar = quotechar + raw_csv->quotechar;
+				gdf->data = stringCol->replace(doublequotechar.c_str(), quotechar.c_str());
+				NVStrings::destroy(stringCol);
+			}
+			else {
+				gdf->data = stringCol;
+			}
 
-		stringColCount++;
-	}
+			RMM_TRY( RMM_FREE( h_str_cols [stringColCount], 0 ) );
 
+			stringColCount++;
+		}
 
-	CUDA_TRY( cudaMemcpy(h_valid_count,d_valid_count, sizeof(unsigned long long) * (raw_csv->num_active_cols), cudaMemcpyDeviceToHost));
 
-	//--- set the null count
-	for ( int col = 0; col < raw_csv->num_active_cols; col++) {
-		cols[col]->null_count = raw_csv->num_records - h_valid_count[col];
-	}
+		CUDA_TRY( cudaMemcpy(h_valid_count,d_valid_count, sizeof(unsigned long long) * (raw_csv->num_active_cols), cudaMemcpyDeviceToHost));
+
+		//--- set the null count
+		for ( int col = 0; col < raw_csv->num_active_cols; col++) {
+			cols[col]->null_count = raw_csv->num_records - h_valid_count[col];
+		}
 
-	free(h_valid_count); 
+		free(h_valid_count); 
```

Comment:
Manual `free`s also give me pause. Is it possible to use a `std::vector` or `std::unique_ptr` instead so you don't have to worry about manually freeing the memory?

Reply:
Please note that I did not add the code here, just put it in a block. Replaced the raw malloc/free with a vector in this case, following the "Boy Scout Rule".
There are many other potential memory leaks in read_csv. We will address these in a future release.